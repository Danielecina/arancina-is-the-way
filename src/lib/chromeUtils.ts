import {ChromeMessage} from '../types'

const queryInfo = {active: true};
export const getCurrentTabUrl = (callback: (url: string | undefined) => void): void => {
  chrome.tabs && chrome.tabs.query(queryInfo, tabs => {
    callback(tabs[0].url);
  });
}

export const getCurrentTabUId = (callback: (url: number | undefined) => void): void => {
  chrome.tabs && chrome.tabs.query(queryInfo, tabs => {
    const tabId = tabs[0] || {}
    callback(tabId.id);
  });
}

export const sendMessage = (message: ChromeMessage) => {
  let response
  getCurrentTabUId(tabId => {
    console.log('sendMessage tabId', tabId)
    tabId && chrome.tabs.sendMessage(
      tabId,
      message,
      {},
      (data) => (response = data)
    )
  })
  return response
}


