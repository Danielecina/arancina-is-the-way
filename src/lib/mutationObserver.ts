import wordReplacerAlgorithm from './algorithm'

export let listener: MutationObserver

export function unsubscribeListener() {
  listener && listener.disconnect()
}

const MUTATION_OBSERVER_CHARACTER_DATA = 'characterData'
const MUTATION_OBSERVER_CHILD_LIST = 'childList'
export function mutationRecord (mutation: MutationRecord) {
  if (mutation.type === MUTATION_OBSERVER_CHARACTER_DATA) {
    wordReplacerAlgorithm(mutation.target)
  } else if (mutation.type === MUTATION_OBSERVER_CHILD_LIST) {
    // monitor the target node (and, if subtree is true, its descendants)
    // for the addition of new child nodes or removal of existing child nodes
    const changedNode = mutation.addedNodes[0]
    wordReplacerAlgorithm(changedNode)
  }
}

export function createListener() {
  listener = new MutationObserver((mutations: MutationRecord[]) => {
    mutations.forEach(mutationRecord)
  })
  
  listener.observe(document.body, {
    childList: true,
    characterData: true,
    subtree: true
  })
}

export default function mutationObserver(watchMode: boolean): undefined {
  if (!watchMode) {
    unsubscribeListener()
    return
  }
  
  wordReplacerAlgorithm()
  if (listener) return
  createListener()
}
