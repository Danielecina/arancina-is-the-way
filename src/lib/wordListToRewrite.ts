interface WrongMap {
  wrongWord: string
  correctWord: string
}

const wordListToRewrite : Array<WrongMap> = [
  {
    wrongWord: 'arancino',
    correctWord: 'arancina'
  },
  {
    wrongWord: 'arancini',
    correctWord: 'arancine'
  },
  {
    wrongWord: 'Arancino',
    correctWord: 'Arancine'
  },
  {
    wrongWord: 'Arancini',
    correctWord: 'Arancine'
  }
]

export default wordListToRewrite
