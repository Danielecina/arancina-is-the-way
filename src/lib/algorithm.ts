import wordListToRewrite from './wordListToRewrite'
import nodeTypeToEvaluate from "./nodeTypeToEvaluate"

function findElementsWithWrongWordAndReplace (element: Node | HTMLElement) {
  if (!element) return
  element.childNodes.forEach(findElementsWithWrongWordAndReplace)
  
  if (
    !element.textContent &&
    !nodeTypeToEvaluate(element.nodeType, element.nodeName)
  ) return
  
  wordListToRewrite.forEach(({wrongWord, correctWord}) => {
    const regex = new RegExp(wrongWord, 'g')
    const matches = (element.textContent || '').match(regex)
    if (matches) {
      element.textContent = (element.textContent || '').replace(regex, correctWord)
    }
  })
}

export default function algorithm (element?: Node)  {
  if (element) {
    findElementsWithWrongWordAndReplace(element)
    return
  }
  
  const rootElement = document.querySelector('body')
  if (!rootElement) return
  
  findElementsWithWrongWordAndReplace(rootElement)
}
