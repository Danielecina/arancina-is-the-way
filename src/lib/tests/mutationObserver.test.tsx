import React from 'react'

import mutationObserver, {listener} from '../mutationObserver'
import * as algorithm from '../algorithm'
import * as chromeUtils from "../chromeUtils"
import wordListToRewrite from "../wordListToRewrite"

jest.spyOn(chromeUtils, 'getCurrentTabUId').mockImplementation(() => Promise.resolve(1))

const mockedAlgorithmResult = jest.fn(() => Promise.resolve({}))
jest.spyOn(algorithm, 'default').mockImplementation(mockedAlgorithmResult)

describe('mutationObserver', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  afterEach(() => {
    listener && listener.disconnect()
  })

  test('expect to register listener correctly', async () => {
    expect(listener).toBeFalsy()
    await mutationObserver(true)
    expect(listener).toBeTruthy()
  })
  
  test('expect to execute algorithm when DOM change', async () => {
    const node = document.createElement('div')
    await mutationObserver(true, node)
  
    expect(listener).toBeTruthy()
    const child = document.createElement('div')
    child.textContent = wordListToRewrite[0].wrongWord
    
    node.appendChild(child)
    expect(node).toMatchSnapshot()
    expect(mockedAlgorithmResult).toHaveBeenCalledWith()
  })
})
