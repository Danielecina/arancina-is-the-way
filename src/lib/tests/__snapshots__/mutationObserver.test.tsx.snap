// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutationObserver expect to execute algorithm when DOM change 1`] = `
<div>
  <div>
    arancino
  </div>
</div>
`;
